d2/src/main/java/com/linkedin/d2/balancer/clients/BackupRequestsClient.java

```diff
@@ -195,24 +247,65 @@ public void restRequest(final RestRequest request, final RequestContext requestC
    */
   private void updateIfNeeded(String serviceName)
   {
-    List<Map<String, Object>> existing = _configs.get(serviceName);
     try
     {
       ServiceProperties serviceProperties = _loadBalancer.getLoadBalancedServiceProperties(serviceName);
-      if (serviceProperties != null)
-      {
-        if (existing != serviceProperties.getBackupRequests())
-        { // reference inequality check
-          update(serviceName, serviceProperties.getBackupRequests());
-          _configs.put(serviceName, serviceProperties.getBackupRequests());
-        }
-      }
-    } catch (ServiceUnavailableException e)
+      updateServiceProperties(serviceName, serviceProperties);
+    }
+    catch (ServiceUnavailableException e)
     {
       LOG.debug("Failed to fetch backup requests strategy ", e);
     }
   }
 
+  private void updateServiceProperties(String serviceName, ServiceProperties serviceProperties)
+  {
+    List<Map<String, Object>> existing = _configs.get(serviceName);
+    if (serviceProperties != null)
+    {
+      if (existing != serviceProperties.getBackupRequests())
+      { // reference inequality check
+        update(serviceName, serviceProperties.getBackupRequests());
+        _configs.put(serviceName, serviceProperties.getBackupRequests());
+      }
+    }
+  }
+
+  void getStrategy(final String serviceName, final String operation, Callback<Optional<TrackingBackupRequestsStrategy>> callback) {
```

Comment:
We now have 2 copies of getStrategy (one sync one async) with almost identical functionality. Can we just keep the async one? You can always wait the results if sync operation is needed.

If necessary, just keep requestAsync and let request to call into requestAsync.

Reply:
I refactored the code a little bit more for better readability. Yes I have 2 copies of the same logic, because I want to be more cautious and reduce the risk. Most of the common logics have been extracted from the 2 copies though. When we want to clean up the blocking call later, it should be fairly easy to clean them up